### PR TITLE
feat: Add and use email verification claim

### PR DESCRIPTION
**Pull request by KShivendu**
## Summary of change

Add and use email verification claim

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `supertokens_python/constants.py`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `setup.py`
    -   In `supertokens_python/constants.py`
-   [ ] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `supertokens_python/utils.py` file to include that in the `FRAMEWORKS` variable
-   [ ] If added a new recipe that has a User type with extra info, then be sure to change the User type in supertokens_python/types.py

### CODE REVIEW

File: supertokens_python/recipe/emailverification/interfaces.py [hunk 1]
```diff
@@ -161,24 +164,55 @@ def __init__(self):
 
     @abstractmethod
     async def email_verify_post(
-        self, token: str, api_options: APIOptions, user_context: Dict[str, Any]
+        self,
+        token: str,
+        api_options: APIOptions,
+        user_context: Dict[str, Any],
+        session: Optional[SessionContainer] = None,
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/interfaces.py:171]: Add session before user_context. And there is no default value for session, but it is optional.

File: supertokens_python/recipe/emailverification/interfaces.py [hunk 2]
```diff
@@ -161,24 +164,55 @@ def __init__(self):
...
     ) -> Union[IsEmailVerifiedGetOkResult, GeneralErrorResponse]:
         pass
 
     @abstractmethod
     async def generate_email_verify_token_post(
-        self, api_options: APIOptions, user_context: Dict[str, Any]
+        self,
+        api_options: APIOptions,
+        user_context: Dict[str, Any],
+        session: SessionContainer,
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/interfaces.py:191]: Add session before user_context

File: supertokens_python/recipe/emailverification/interfaces.py [hunk 3]
```diff
@@ -161,24 +164,55 @@ def __init__(self):
...
     ]:
         pass
 
     @abstractmethod
     async def is_email_verified_get(
-        self, api_options: APIOptions, user_context: Dict[str, Any]
+        self,
+        api_options: APIOptions,
+        user_context: Dict[str, Any],
+        session: Optional[SessionContainer] = None,
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/interfaces.py:182]: Add session before user_context

File: supertokens_python/recipe/emailverification/interfaces.py [hunk 4]
```diff
@@ -121,8 +125,7 @@ def to_json(self) -> Dict[str, Any]:
 
 
 class EmailVerifyPostInvalidTokenError(APIResponse):
-    def __init__(self):
-        self.status = "EMAIL_VERIFICATION_INVALID_TOKEN_ERROR"
+    status = "EMAIL_VERIFICATION_INVALID_TOKEN_ERROR"
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/interfaces.py:128]: why this change?

File: supertokens_python/recipe/emailverification/api/email_verify.py [hunk 5]
```diff
@@ -40,18 +43,29 @@ async def handle_email_verify_api(
...
         if api_implementation.disable_is_email_verified_get:
             return None
 
-        user_context = default_user_context(api_options.request)
+        session = await get_session(
+            api_options.request,
+            override_global_claim_validators=lambda _, __, ___: [],
+            user_context=user_context,
+        )
+
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/api/email_verify.py:66]: how come up do not have to do `assert session is not None`? Cause you do that in generate_email_verify_token.py` file.

File: supertokens_python/recipe/emailverification/__init__.py [hunk 6]
```diff
@@ -13,26 +13,45 @@
...
     from supertokens_python.supertokens import AppInfo
 
     from ...recipe_module import RecipeModule
 
+from .utils import MODE_TYPE, OverrideConfig
+
 
 def init(
-    config: ParentRecipeEmailVerificationConfig,
+    mode: MODE_TYPE = "OPTIONAL",
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/__init__.py:43]: there is no default value to this. The user must pass some value themselves.

File: supertokens_python/recipe/emailverification/utils.py [hunk 7]
```diff
@@ -13,8 +13,8 @@
 # under the License.
 
 from __future__ import annotations
-
-from typing import TYPE_CHECKING, Any, Dict
+from typing_extensions import Literal
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/utils.py:16]: will this work in python 3.7 and greater?

File: supertokens_python/recipe/emailverification/utils.py [hunk 8]
```diff
@@ -57,26 +44,26 @@ def __init__(
         self.apis = apis
 
 
+MODE_TYPE = Literal["REQUIRED", "OPTIONAL"]
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/utils.py:47]: does this do a compile error if the user doesn't give one of these two (even if they do not use a linter). Cause if not, we should have input validation to make sure that the value is indeed one of these and if not, throw an appropriate error.

File: supertokens_python/recipe/emailverification/utils.py [hunk 9]
```diff
@@ -57,26 +44,26 @@ def __init__(
...
+
+
 class ParentRecipeEmailVerificationConfig:
+    # TODO: Rename this? ^
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/utils.py:51]: yea.. we should. Since now no other recipe embeds this one.

File: supertokens_python/recipe/emailverification/utils.py [hunk 10]
```diff
@@ -57,26 +44,26 @@ def __init__(
...
     def __init__(
         self,
-        get_email_for_user_id: Callable[[str, Dict[str, Any]], Awaitable[str]],
-        override: Union[OverrideConfig, None] = None,
-        get_email_verification_url: Union[
-            Callable[[User, Dict[str, Any]], Awaitable[str]], None
-        ] = None,
+        mode: MODE_TYPE = "OPTIONAL",
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/utils.py:54]: no default value for mode

File: supertokens_python/recipe/emailverification/utils.py (hunk 9 above)
Comment (rishabhpoddar) [reply supertokens_python/recipe/emailverification/utils.py:51]: Actually, this class is not really needed anymore.. you can just pass in the individual arguments to the `validate_and_normalise_user_input` function (just like in other recipes)

File: supertokens_python/recipe/emailverification/api/implementation.py [hunk 11]
```diff
@@ -13,105 +13,7 @@
...
-    EmailVerifyPostOkResult,
-    GenerateEmailVerifyTokenPostEmailAlreadyVerifiedError,
-    GenerateEmailVerifyTokenPostOkResult,
-    IsEmailVerifiedGetOkResult,
-    VerifyEmailUsingTokenOkResult,
-)
+from typing import TYPE_CHECKING
 
 if TYPE_CHECKING:
-    from supertokens_python.recipe.emailverification.interfaces import APIOptions
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/api/implementation.py:31]: why is this file empty? Most of the stuff you removed is needed..

File: supertokens_python/recipe/emailverification/asyncio/__init__.py [hunk 12]
```diff
@@ -37,20 +55,73 @@ async def verify_email_using_token(
...
+
+    recipe = EmailVerificationRecipe.get_instance()
+    if email is None:
+        email_info = await recipe.get_email_for_user_id(user_id, user_context)
+        if isinstance(email_info, GetEmailForUserIdOkResult):
+            email = email_info.email
+        elif isinstance(email_info, EmailDoesnotExistError):
+            # Here we are returning OK since that's how it used to work, but a later call
+            # to is_verified will still return true
+            return CreateEmailVerificationTokenEmailAlreadyVerifiedError()
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/asyncio/__init__.py:96]: this should return `RevokeEmailVerificationTokensOkResult`. This is also a mistake in the node PR, and I have commented in that.

Comment (rishabhpoddar) [reply supertokens_python/recipe/emailverification/asyncio/__init__.py:96]: Also the comment you added vs the code you wrote.. copy / pasting error?

File: supertokens_python/recipe/emailverification/recipe.py [hunk 13]
```diff
@@ -161,16 +184,43 @@ def get_all_cors_headers(self) -> List[str]:
         return []
 
     @staticmethod
-    def init(config: ParentRecipeEmailVerificationConfig):
+    def init(
+        mode: MODE_TYPE = "OPTIONAL",
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/recipe.py:188]: no default value for mode

File: supertokens_python/recipe/emailverification/recipe.py [hunk 14]
```diff
@@ -186,10 +236,164 @@ def get_instance() -> EmailVerificationRecipe:
...
+
+        super().__init__("st-ev", fetch_value)
+
+        self.validators = EmailVerificationClaimValidators(claim=self)
+
+
+EmailVerificationClaim = EmailVerificationClaimClass()
+
+
+class APIImplementation(APIInterface):
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/recipe.py:295]: why is this moved into this file and not kept in the other file?

File: supertokens_python/recipe/emailverification/recipe.py [hunk 15]
```diff
@@ -186,10 +236,164 @@ def get_instance() -> EmailVerificationRecipe:
...
+            if not isinstance(res, UnknownUserIdError):
+                return res
+
+        return UnknownUserIdError()
+
+    def add_get_email_for_user_id_func(self, f: Callable[[str, Dict[str, Any]], Any]):
+        self.get_email_for_user_id_funcs_from_other_recipes.append(f)
+
+
+class EmailVerificationClaimClass(BooleanClaim):
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/recipe.py:270]: why is this class in this file and not in the ev_claim.py file?

File: supertokens_python/recipe/emailverification/recipe.py [hunk 16]
```diff
@@ -186,10 +236,164 @@ def get_instance() -> EmailVerificationRecipe:
...
+                return res
+
+        for f in self.get_email_for_user_id_funcs_from_other_recipes:
+            res = await f(user_id, user_context)
+            if not isinstance(res, UnknownUserIdError):
+                return res
+
+        return UnknownUserIdError()
+
+    def add_get_email_for_user_id_func(self, f: Callable[[str, Dict[str, Any]], Any]):
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/recipe.py:266]: why is the `f` argument so badly typed? Why is the return of it `Any`?

File: supertokens_python/recipe/emailverification/recipe.py [hunk 17]
```diff
@@ -186,10 +236,164 @@ def get_instance() -> EmailVerificationRecipe:
...
+            raise Exception(
+                "Should never come here: UNKNOWN_USER_ID or invalid result from get_email_for_user"
+            )
+
+        super().__init__("st-ev", fetch_value)
+
+        self.validators = EmailVerificationClaimValidators(claim=self)
+
+
+EmailVerificationClaim = EmailVerificationClaimClass()
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/recipe.py:292]: how will the end user import this in their code (if they want to)?

File: supertokens_python/recipe/emailverification/recipe.py [hunk 18]
```diff
@@ -186,10 +236,164 @@ def get_instance() -> EmailVerificationRecipe:
...
+        user_context: Dict[str, Any],
+        session: Optional[SessionContainer] = None,
+    ) -> IsEmailVerifiedGetOkResult:
+        if session is None:
+            raise Exception("Session is undefined. Should not come here.")
+        await session.fetch_and_set_claim(EmailVerificationClaim, user_context)
+        is_verified = await session.get_claim_value(
+            EmailVerificationClaim, user_context
+        )
+        # TODO: Type of is_verified should be bool. It's any for now.
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailverification/recipe.py:326]: TODO

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: supertokens_python/recipe/emailpassword/emaildelivery/services/smtp/service_implementation/__init__.py [hunk 19]
```diff
@@ -45,14 +41,6 @@ def __init__(self, transporter: Transporter) -> None:
...
         self.ev_get_content = email_verification_service_implementation.get_content
 
-        derived_ev_service_implementation = DerivedEVServiceImplementation(self)
-        email_verification_service_implementation.send_raw_email = (
-            derived_ev_service_implementation.send_raw_email
-        )
-        email_verification_service_implementation.get_content = (
-            derived_ev_service_implementation.get_content
-        )
-
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/emailpassword/emaildelivery/services/smtp/service_implementation/__init__.py:55]: Fixed.

File: supertokens_python/recipe/emailpassword/api/implementation.py [hunk 20]
```diff
@@ -92,9 +92,8 @@ async def generate_password_reset_token_post(
 
         token = token_result.token
         password_reset_link = (
-            await api_options.config.reset_password_using_token_feature.get_reset_password_url(
-                user, user_context
-            )
+            api_options.app_info.website_domain.get_as_string_dangerous()
+            + api_options.app_info.website_base_path.get_as_string_dangerous()
             + "?token="
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/emailpassword/api/implementation.py:97]: Fixed.

File: supertokens_python/recipe/emailpassword/utils.py [hunk 21]
```diff
@@ -380,19 +322,18 @@ def __init__(
         self.sign_up_feature = sign_up_feature
         self.sign_in_feature = sign_in_feature
         self.reset_password_using_token_feature = reset_password_using_token_feature
-        self.email_verification_feature = email_verification_feature
+        # self.email_verification_feature = email_verification_feature
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/emailpassword/utils.py:325]: Fixed.

File: supertokens_python/recipe/emailpassword/utils.py [hunk 22]
```diff
@@ -371,7 +313,7 @@ def __init__(
         sign_up_feature: SignUpFeature,
         sign_in_feature: SignInFeature,
         reset_password_using_token_feature: ResetPasswordUsingTokenFeature,
-        email_verification_feature: ParentRecipeEmailVerificationConfig,
+        # email_verification_feature: ParentRecipeEmailVerificationConfig,
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/emailpassword/utils.py:316]: Fixed.

File: supertokens_python/recipe/emailpassword/utils.py [hunk 23]
```diff
@@ -224,36 +202,33 @@ def __init__(
...
+#             Callable[[User, str, Dict[str, Any]], Awaitable[None]], None
+#         ] = None,
+#     ):
+#         self.get_email_verification_url = get_email_verification_url
+#         self.create_and_send_custom_email = create_and_send_custom_email
+#
+#         if create_and_send_custom_email:
+#             deprecated_warn(
+#                 "create_and_send_custom_email is deprecated. Please use email delivery config instead"
+#             )
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/emailpassword/utils.py:227]: Fixed.

File: supertokens_python/recipe/emailpassword/recipe.py [hunk 24]
```diff
@@ -23,17 +23,18 @@
...
 )
-from supertokens_python.recipe.emailverification.types import (
-    EmailVerificationIngredients,
-    VerificationEmailTemplateVars,
-)
 from supertokens_python.recipe_module import APIHandled, RecipeModule
+from ..emailverification.interfaces import (
+    UnknownUserIdError,
+    GetEmailForUserIdOkResult,
+    EmailDoesnotExistError,
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/emailpassword/recipe.py:30]: Fixed.

File: supertokens_python/recipe/thirdparty/api/implementation.py [hunk 25]
```diff
@@ -162,14 +163,18 @@ async def sign_in_up_post(
         )
 
         if email_verified:
-            token_response = await api_options.email_verification_recipe_implementation.create_email_verification_token(
-                user_id=signinup_response.user.user_id,
-                email=signinup_response.user.email,
-                user_context=user_context,
+            ev_instance = EmailVerificationRecipe.get_instance()
+            assert ev_instance is not None
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/thirdparty/api/implementation.py:167]: TODO

File: supertokens_python/recipe/thirdparty/api/implementation.py [hunk 26]
```diff
@@ -162,14 +163,18 @@ async def sign_in_up_post(
         )
 
         if email_verified:
-            token_response = await api_options.email_verification_recipe_implementation.create_email_verification_token(
-                user_id=signinup_response.user.user_id,
-                email=signinup_response.user.email,
-                user_context=user_context,
+            ev_instance = EmailVerificationRecipe.get_instance()
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/thirdparty/api/implementation.py:166]: Done.

File: supertokens_python/recipe/thirdparty/recipe.py [hunk 27]
```diff
@@ -168,38 +138,31 @@ async def handle_api_request(
...
-        )
+        return None  # TODO: Node PR returns False, but here signature is different. Verify if this is correct.
 
     async def handle_error(
         self, request: BaseRequest, err: SuperTokensError, response: BaseResponse
-    ) -> BaseResponse:
+    ) -> BaseResponse:  # type: ignore
         if isinstance(err, SuperTokensThirdPartyError):
             raise err
-        return await self.email_verification_recipe.handle_error(request, err, response)
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/thirdparty/recipe.py:180]: > just raise err

I did. But pyright didn't allow me to do that so I had to use `type: ignore`. Do you know anything better?

File: supertokens_python/recipe/thirdpartyemailpassword/recipe.py [hunk 28]
```diff
@@ -346,12 +297,11 @@ def reset():
             raise Exception(None, "calling testing function in non testing env")
         ThirdPartyEmailPasswordRecipe.__instance = None
 
-    async def get_email_for_user_id(
-        self, user_id: str, user_context: Dict[str, Any]
-    ) -> str:
+    async def get_email_for_user_id(self, user_id: str, user_context: Dict[str, Any]):
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/thirdpartyemailpassword/recipe.py:300]: Okay. Removed.

File: supertokens_python/recipe/passwordless/api/implementation.py [hunk 29]
```diff
@@ -272,9 +271,27 @@ async def consume_code_post(
             return ConsumeCodePostRestartFlowError()
 
         user = response.user
+
+        if user.email is not None:
+            ev_instance = EmailVerificationRecipe.get_instance()
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/passwordless/api/implementation.py:276]: Fixed.

File: supertokens_python/recipe/passwordless/recipe.py [hunk 30]
```diff
@@ -202,7 +217,7 @@ async def handle_api_request(
 
     async def handle_error(
         self, request: BaseRequest, err: SuperTokensError, response: BaseResponse
-    ):
+    ) -> BaseResponse:  # type: ignore
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/passwordless/recipe.py:220]: I tried. But pylint doesn't let me just raise the error. It needs a return value. Do you know anything better?

File: supertokens_python/recipe/emailverification/recipe.py (hunk 14 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailverification/recipe.py:295]: There was a cyclic import error that kept me stuck for hours. I still don't know how to fix it. Generally, this happens because of types so we just use `from _future__  import annotations` and `from typing import TYPE_CHECKING`.

But here we are using using the class (not just its type) so the error is really hard to get rid of. I have struggled with it for hours. The only feasible solution for now was to move it here. Need some help to fix it.

File: supertokens_python/recipe/emailpassword/utils.py [hunk 31]
```diff
@@ -405,10 +346,10 @@ def validate_and_normalise_user_input(
...
         )
 
-    if email_verification_feature is not None and not isinstance(email_verification_feature, InputEmailVerificationConfig):  # type: ignore
-        raise ValueError(
-            "email_verification_feature must be of type InputEmailVerificationConfig or None"
-        )
+    # if email_verification_feature is not None and not isinstance(email_verification_feature, InputEmailVerificationConfig):  # type: ignore
+    #     raise ValueError(
+    #         "email_verification_feature must be of type InputEmailVerificationConfig or None"
+    #     )
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/emailpassword/utils.py:352]: Removed this. Looked for more using Ctrl+F didn't find any others.

File: supertokens_python/recipe/passwordless/interfaces.py [hunk 32]
```diff
@@ -31,6 +31,7 @@
     User,
 )
 from .utils import PasswordlessConfig
+from ...supertokens import AppInfo
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/passwordless/interfaces.py:35]: Done. Thanks!

File: supertokens_python/recipe/thirdpartypasswordless/utils.py [hunk 33]
```diff
@@ -115,32 +93,6 @@ async def func(user: EmailVerificationUser, user_context: Dict[str, Any]):
...
-
-    return ParentRecipeEmailVerificationConfig(
-        get_email_for_user_id=recipe.get_email_for_user_id,
-        create_and_send_custom_email=create_and_send_custom_email,
-        get_email_verification_url=get_email_verification_url,
-        override=override.email_verification_feature,
-    )
-
-
 class InputOverrideConfig:
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/thirdpartypasswordless/utils.py:96]: Done. Verified the same using Ctrl+F for `InputOverrideConfig:`

File: supertokens_python/recipe/thirdpartypasswordless/utils.py [hunk 34]
```diff
@@ -54,33 +53,12 @@
...
-    ):
-        self.get_email_verification_url = get_email_verification_url
-        self.create_and_send_custom_email = create_and_send_custom_email
-        if create_and_send_custom_email:
-            deprecated_warn(
-                "create_and_send_custom_email is deprecated. Please use email delivery config instead"
-            )
-
-
 def email_verification_create_and_send_custom_email(
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/thirdpartypasswordless/utils.py:62]: Done and searched for "verification" too.

File: supertokens_python/recipe/thirdpartypasswordless/utils.py [hunk 35]
```diff
@@ -54,33 +53,12 @@
...
     recipe: ThirdPartyPasswordlessRecipe,
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/thirdpartypasswordless/utils.py:63]: Done.

File: supertokens_python/recipe/thirdparty/api/implementation.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/thirdparty/api/implementation.py:167]: done.

File: supertokens_python/recipe/session/asyncio/__init__.py [hunk 36]
```diff
@@ -13,6 +13,8 @@
 # under the License.
 from typing import Any, Dict, List, Union, TypeVar, Callable, Optional
 
+# TODO: Finalize how to import Optional
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/session/asyncio/__init__.py:16]: No problem with python3.7+. Nothing needs to change.

File: supertokens_python/recipe/session/asyncio/__init__.py [hunk 37]
```diff
@@ -37,6 +39,8 @@
     GetJWKSResult,
 )
 
+# TODO: https://github.com/supertokens/supertokens-python/pull/209#discussion_r932049999
+# There have been changes to this function as well. See this:
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/session/asyncio/__init__.py:43]: Done. https://github.com/supertokens/supertokens-python/pull/209#discussion_r939839717

File: supertokens_python/recipe/thirdpartypasswordless/recipe.py [hunk 38]
```diff
@@ -418,7 +298,6 @@ def func(app_info: AppInfo):
                     ingredients,
                     get_link_domain_and_path,
                     get_custom_user_input_code,
-                    email_verification_feature,
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/thirdpartypasswordless/recipe.py:421]: Removed.

File: supertokens_python/recipe/session/recipe.py [hunk 39]
```diff
@@ -302,6 +302,9 @@ def reset():
         ):
             raise_general_exception("calling testing function in non testing env")
         SessionRecipe.__instance = None
+        # FIXME: Discovered its requirement while running tests. Confirm if this is correct:
+        SessionRecipe.claims_added_by_other_recipes = []
+        SessionRecipe.claim_validators_added_by_other_recipes = []
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/session/recipe.py]: Done.

File: supertokens_python/recipe/session/interfaces.py [hunk 40]
```diff
@@ -604,5 +605,8 @@ async def validate(
     def should_refetch(  # pylint: disable=no-self-use
         self, payload: JSONObject, user_context: Dict[str, Any]
     ) -> MaybeAwaitable[bool]:
+        # TODO: https://github.com/supertokens/supertokens-python/pull/209#discussion_r932121943
+        # TODO: This should also be an abstractmethod
+        # TODO: This should also be async
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/session/interfaces.py]: Done.

Final implementation is based on what we discussed in the base PR.

File: supertokens_python/recipe/session/claim_base_classes/primitive_claim.py [hunk 41]
```diff
@@ -171,6 +172,9 @@ def add_to_payload_(
     def remove_from_payload_by_merge_(
         self, payload: JSONObject, user_context: Dict[str, Any]
     ) -> JSONObject:
+        # TODO: https://github.com/supertokens/supertokens-python/pull/209#discussion_r931922854
+        # Not sure if this will actually work in python -> cause there is no diff between "null" and
+        # "undefined" in python (both are None). So setting this to None is as good as deleting it from the map?
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/session/claim_base_classes/primitive_claim.py:179]: Yes. Removed comment.

File: supertokens_python/recipe/session/claim_base_classes/primitive_claim.py [hunk 42]
```diff
@@ -127,6 +127,7 @@ async def validate(
 
 
 class PrimitiveClaimValidators(Generic[_T]):
+    # TODO: We should discuss how someone would override existing validators
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply supertokens_python/recipe/session/claim_base_classes/primitive_claim.py:131]: Done.

File: supertokens_python/recipe/emailverification/interfaces.py (hunk 1 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailverification/interfaces.py:171]: Done.

File: supertokens_python/recipe/emailverification/interfaces.py (hunk 2 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailverification/interfaces.py:191]: Done.

File: supertokens_python/recipe/emailverification/interfaces.py (hunk 4 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailverification/interfaces.py:128]: Hmm. it's not required. 

Restored it.

File: supertokens_python/recipe/emailverification/api/email_verify.py (hunk 5 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailverification/api/email_verify.py:66]: TODO

File: supertokens_python/recipe/emailverification/__init__.py (hunk 6 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailverification/__init__.py:43]: Okay. Removed the default value.

Done.

File: supertokens_python/recipe/emailverification/utils.py (hunk 8 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailverification/utils.py:47]: It doesn't give compile error. 

Added validation in `validate_and_normalise_user_input` of the ev recipe. Error message would be `Email Verification recipe mode must be one of 'REQUIRED' or 'OPTIONAL'`.

File: supertokens_python/recipe/emailverification/utils.py (hunk 9 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailverification/utils.py:51]: TODO.

File: supertokens_python/recipe/emailverification/utils.py (hunk 10 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailverification/utils.py:54]: Done.

File: supertokens_python/recipe/emailverification/api/implementation.py (hunk 11 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailverification/api/implementation.py:31]: Didn't remove anything. I had to move the content to avoid cyclic error. It's really troublesome. Have already wasted hours. Need help to fix it.

File: supertokens_python/recipe/emailverification/recipe.py (hunk 13 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailverification/recipe.py:188]: Removed.

File: supertokens_python/recipe/emailverification/recipe.py (hunk 15 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailverification/recipe.py:270]: I had to move the content to avoid cyclic import errors. It's really troublesome this time. Have already wasted hours. Need help to fix it.

File: supertokens_python/recipe/emailverification/recipe.py (hunk 16 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailverification/recipe.py:266]: Fixed.

File: supertokens_python/recipe/emailverification/recipe.py (hunk 17 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailverification/recipe.py:292]: `from supertokens_python.recipe.emailverification import EmailVerificationClaim` 

I know `EmailVerificationClaim` and `EmailVerificationClaimClass()` shouldn't be in `recipe.py`. But there's some nasty cyclic import errors arise if I seperate them. Need help in fixing them.

File: supertokens_python/recipe/emailverification/utils.py (hunk 9 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailverification/utils.py]: Removed ParentRecipeEmailVerificationConfig

Done.
